include/autoppl/expr_builder.hpp

```diff
@@ -0,0 +1,164 @@
+                    >>
+{
+    using type = T;
+};
+
+template <class T>
+using convert_to_cont_dist_param_t = 
+    typename convert_to_cont_dist_param<T>::type;
+
+} // namespace details
+
+#ifndef AUTOPPL_USE_CONCEPTS
+/*
+ * Builds a Uniform expression only when the parameters
+ * are both valid continuous distribution parameter types.
+ * See var_expr.hpp for more information.
+ */
+template <class MinType, class MaxType>
+inline constexpr auto uniform(MinType&& min_expr,
+                              MaxType&& max_expr)
+{
+    using min_t = details::convert_to_cont_dist_param_t<MinType>;
+    using max_t = details::convert_to_cont_dist_param_t<MaxType>;
+
+    min_t wrap_min_expr = std::forward<MinType>(min_expr);
+    max_t wrap_max_expr = std::forward<MaxType>(max_expr);
+
+    return expr::Uniform(wrap_min_expr, wrap_max_expr);
+}
+#else
+#endif
+
+#ifndef AUTOPPL_USE_CONCEPTS
+/*
+ * Builds a Normal expression only when the parameters
+ * are both valid continuous distribution parameter types.
+ * See var_expr.hpp for more information.
+ */
+template <class MeanType, class StddevType>
+inline constexpr auto normal(MeanType&& mean_expr,
+                             StddevType&& stddev_expr)
+{
+    using mean_t = details::convert_to_cont_dist_param_t<MeanType>;
+    using stddev_t = details::convert_to_cont_dist_param_t<StddevType>;
+
+    mean_t wrap_mean_expr = std::forward<MeanType>(mean_expr);
+    stddev_t wrap_stddev_expr = std::forward<StddevType>(stddev_expr);
+
+    return expr::Normal(wrap_mean_expr, wrap_stddev_expr);
+}
+
+#else
+#endif
+
+#ifndef AUTOPPL_USE_CONCEPTS
+/*
+ * Builds a Bernoulli expression only when the parameter
+ * is a valid discrete distribution parameter type.
+ * See var_expr.hpp for more information.
+ * TODO: generalize as done with uniform and normal
+ */
+template <class ProbType>
+inline constexpr auto bernoulli(const ProbType& p_expr)
+{
+    return expr::Bernoulli(p_expr);
+}
+#else
+#endif
+
+////////////////////////////////////////////////////////
+// Model Expression Builder
+////////////////////////////////////////////////////////
+
+#ifndef AUTOPPL_USE_CONCEPTS
+/*
+ * Builds an EqNode to associate var with dist
+ * only when var is a Variable and dist is a valid distribution expression.
+ * Ex. x |= uniform(0,1)
+ */
+template <class T, class DistType>
+inline constexpr auto operator|=(Variable<T>& var,
+                                 DistType&& dist)
+{ return expr::EqNode(var, std::forward<DistType>(dist)); }
+#else
+#endif
+
+#ifndef AUTOPPL_USE_CONCEPTS
+/*
+ * Builds a GlueNode to "glue" the left expression with the right
+ * only when both parameters are valid model expressions.
+ * Ex. (x |= uniform(0,1), y |= uniform(0, 2))
+ */
+template <class LHSNodeType, class RHSNodeType>
+inline constexpr auto operator,(LHSNodeType&& lhs,
+                                RHSNodeType&& rhs)
+{ 
+    return expr::GlueNode(std::forward<LHSNodeType>(lhs), 
```

Comment:
Can you elaborate on what std::forward does and why we always pass to e.g. operator , by rvalue reference? Why does this lvalue/rvalue distinction matter?

Reply:
(repost for record) Since the parameter types are universal references, the deduced types may be any of the 4: const A&, A&, const A&&, A&&. Whatever the deduced type is, we want to pass the arguments into the next function call exactly how they were deduced. If you don't forward, unfortunately the next function call will deduce them to be lvalue references always. Here's something that might perplex you: an rvalue reference itself is an lvalue reference xD